### ISSUE TITLE
Custom substance/trace dynamics

### ISSUE THREAD

**Issue by Mathadon** [iea-annex60/modelica-annex60#390]
Currently we have in `LumpedVolumeDeclarations`:
```
  parameter Modelica.Fluid.Types.Dynamics energyDynamics=Modelica.Fluid.Types.Dynamics.DynamicFreeInitial 
    "Formulation of energy balance"
    annotation(Evaluate=true, Dialog(tab = "Dynamics", group="Equations"));
  parameter Modelica.Fluid.Types.Dynamics massDynamics=energyDynamics 
    "Formulation of mass balance"
    annotation(Evaluate=true, Dialog(tab = "Dynamics", group="Equations"));
  final parameter Modelica.Fluid.Types.Dynamics substanceDynamics=energyDynamics 
    "Formulation of substance balance"
    annotation(Evaluate=true, Dialog(tab = "Dynamics", group="Equations"));
  final parameter Modelica.Fluid.Types.Dynamics traceDynamics=energyDynamics 
    "Formulation of trace substance balance"
    annotation(Evaluate=true, Dialog(tab = "Dynamics", group="Equations"));
```

can we remove the `final`? I want to be able to set steady state trace dynamics and dynamic energyDynamics and this is currently not possible.

**Comment by mwetter** [comment]
In the current implementation, I made it `final` as I did not anticipate any benefit of selecting something else than for the energy balance. Do you really get faster simulation by selecting a different value? If so, then for `traceDynamics` removing the `final` is OK, a use case may be to add a dynamic energy balance to avoid algebraic loops on the energy and mass flow simulation but this may not be needed for the trace dynamics.

For the `substanceDynamics`, I suggest to leave it as it is unless there is a demonstrated benefit for adding complexity by exposing this parameter.

**Comment by Mathadon** [comment]
In some cases the trace substance dynamics are the limiting factor for my Euler time step and since there is no parameter like `mSenFac` I can't play with that. So it would at least be convenient to have this option.

Now that we're looking at these parameters I was thinking if the comment could be made clearer. Currently we have: 
```
"Formulation of energy balance"
"Formulation of mass balance"
"Formulation of substance balance"
```

I was thinking of something like:
```
"Type of energy balance: dynamic (3 initialisation options) or static"
"Type of mass balance: dynamic (3 initialisation options) or static"
"Type of substance balance: dynamic (3 initialisation options) or static"
```

**Comment by mwetter** [comment]
@Mathadon This wording would be good, but please use the word "trace" in `Formulation of trace substance balance`.

 If you do the changes with a script, it would be useful if you can post the script as these parameters are propagated in various models that are not in Annex60, but in Buildings and probably also in IDEAS, AixLib and BuildingSystems.

**Comment by Mathadon** [comment]
I pushed the changes to the existing pull request.

Note that I kept the wording `Type of substance balance` (Medium.Xi) since otherwise there is no distinction with the `Type of trace substance balance` (Medium.C).

The python script:
```
import os

for dname, dirs, files in os.walk("/Users/Werk/Documents/Software/Annex60-git/Annex60"):
    for fname in files:
        fpath = os.path.join(dname, fname)
        with open(fpath) as f:
            s = f.read()
        s = s.replace("\"Formulation of energy balance\"", "\"Type of energy balance: dynamic (3 initialisation options) or static\"")
        s = s.replace("\"Formulation of mass balance\"", "\"Type of mass balance: dynamic (3 initialisation options) or static\"")
        s = s.replace("\"Formulation of substance balance\"", "\"Type of substance balance: dynamic (3 initialisation options) or static\"")
        s = s.replace("\"Formulation of trace substance balance\"", "\"Type of trace substance balance: dynamic (3 initialisation options) or static\"")
        with open(fpath, "w") as f:
            f.write(s)
```
be sure not to include the .git directory as I do not know if the script may screw up that directory.

**Comment by Mathadon** [comment]
I replaced `static` by `steady state` since this wording is more consistent with the `SteadyState` option.

**Comment by mwetter** [comment]
@Mathadon 
I think you changed the wrong parameter. You removed the `final` from `substanceDynamics` but I thought you want to be able to change the dynamic balance for `C`, which is governed by `traceDynamics`.

I also changed the spelling to *initialization* to use consistent spelling within the library and with the MSL, and used *independent mass fraction* instead of *substance* to avoid the confusion with trace substance.

The conversion script is now
```
import os

for dname, dirs, files in os.walk("../Annex60"):
    for fname in files:
        fpath = os.path.join(dname, fname)
        with open(fpath) as f:
            s = f.read()
        s = s.replace("\"Formulation of energy balance\"", \
            "\"Type of energy balance: dynamic (3 initialisation options) or steady state\"")
        s = s.replace("\"Formulation of mass balance\"", \
            "\"Type of mass balance: dynamic (3 initialization options) or steady state\"")
        s = s.replace("\"Formulation of substance balance\"", \
            "\"Type of independent mass fraction balance: dynamic (3 initialization options) or steady state\"")
        s = s.replace("\"Formulation of trace substance balance\"", \
            "\"Type of trace substance balance: dynamic (3 initialization options) or steady state\"")

        with open(fpath, "w") as f:
            f.write(s)
```

What is still open is

- [ ] Correct the wrong parameter (expose `traceDynamics`, not `substanceDynamics`)
- [ ] Update the `Fluid.UsersGuide`

**Comment by Mathadon** [comment]
I did intend to change the `substanceDynamics`. It turns out that I can do without the option right now since the model becomes slower, so if you're still rather against the option we could drop it for now, but I think it's nice to have. :)

**Comment by mwetter** [comment]
I am more in favor for not exposing `substanceDynamics` and rather use `final substanceDynamics = energyDynamics` as we have now. The justification for this is that if a volume is large enough to warrant having a dynamic energy balance, then the moisture that is brought into this volume by the air flow should also be dynamic, and vice versa.

In fact, during the revisions, I was thinking why not eliminating `substanceDynamics` and use instead `energyDynamics`. Then, we would have `energyDynamics`, `massDynamics` and `traceDynamics`.

**Comment by Mathadon** [comment]
It makes sense that `substanceDynamics` and `energyDynamics` are closely related, but that is what we have the default value for. In my opinion the models should be reconfigurable such that when someone (for god knows what reason) needs to change the the dynamics individually, this is possible. But I understand that having another dynamics parameter may be confusing to some users.. I would therefore propose to revert my changes and stick with the existing implementation since we currently do not have a good motivation to make a change and then this is the least intrusive option.

**Comment by mwetter** [comment]
I will revert the changes based on the web conference discussion.

**Comment by mwetter** [comment]
I made a new branch `issue390_substancedynamics_refactored` to revert the changes. This is based on the current `master`.

For c72c807, I applied the following script:
```
import os

for dname, dirs, files in os.walk("../Annex60"):
    for fname in files:
        fpath = os.path.join(dname, fname)
        with open(fpath) as f:
            s = f.read()
        s = s.replace("\"Formulation of energy balance\"", \
            "\"Type of energy balance: dynamic (3 initialization options) or steady state\"")
        s = s.replace("\"Formulation of mass balance\"", \
            "\"Type of mass balance: dynamic (3 initialization options) or steady state\"")
        s = s.replace("\"Formulation of substance balance\"", \
            "\"Type of independent mass fraction balance: dynamic (3 initialization options) or steady state\"")
        s = s.replace("\"Formulation of trace substance balance\"", \
            "\"Type of trace substance balance: dynamic (3 initialization options) or steady state\"")

        with open(fpath, "w") as f:
            f.write(s)

```

**Comment by Mathadon** [comment]
Fixed in #421
